FWCore/Framework/src/EventSetupRecordProvider.cc

```diff
@@ -133,97 +119,121 @@ namespace edm {
-        return true;
+      if (intervalStatus_ == IntervalStatus::Invalid || !validityInterval_.validFor(iTime)) {
+        return finder_->nonconcurrentAndIOVNeedsUpdate(key_, iTime);
       }
-      bool returnValue = false;
-      //need to see if we get a new interval
-      if (nullptr != finder_.get()) {
-        IOVSyncValue oldFirst(validityInterval_.first());
-
-        validityInterval_ = finder_->findIntervalFor(key_, iTime);
-        //are we in a valid range?
-        if (validityInterval_.first() != IOVSyncValue::invalidIOVSyncValue()) {
-          returnValue = true;
-          //did we actually change?
-          if (oldFirst != validityInterval_.first()) {
-            //tell all Providers to update
-            for (auto& provider : providers_) {
-              provider->newInterval(key_, validityInterval_);
+      return false;
+    }
+
+    bool EventSetupRecordProvider::setValidityIntervalFor(const IOVSyncValue& iTime) {
+      // This function can be called multiple times for the same
+      // IOVSyncValue because DependentRecordIntervalFinder::setIntervalFor
+      // can call it in addition to it being called directly. Don't
+      // need to do the work multiple times for the same call to
+      // eventSetupForInstance.
+      if (intervalStatus_ == IntervalStatus::NotInitializedForSyncValue) {
+        intervalStatus_ = IntervalStatus::Invalid;
+
+        if (validityInterval_.first() != IOVSyncValue::invalidIOVSyncValue() && validityInterval_.validFor(iTime)) {
+          intervalStatus_ = IntervalStatus::SameInterval;
+
+        } else if (finder_.get() != nullptr) {
+          IOVSyncValue oldFirst(validityInterval_.first());
+          IOVSyncValue oldLast(validityInterval_.last());
+          validityInterval_ = finder_->findIntervalFor(key_, iTime);
+
+          // An interval is valid if and only if the start of the interval is
+          // valid. If the start is valid and the end is invalid, it means we
+          // do not know when the interval ends, but the interval is valid and
+          // iTime is within the interval.
+          if (validityInterval_.first() != IOVSyncValue::invalidIOVSyncValue()) {
+            // An interval is new if the start of the interval changes
+            if (validityInterval_.first() != oldFirst) {
+              intervalStatus_ = IntervalStatus::NewInterval;
+
+              // If the start is the same but the end changes, we consider
+              // this the same interval because we do not want to do the
+              // work to update the caches of data in this case.
+            } else if (validityInterval_.last() != oldLast) {
+              intervalStatus_ = IntervalStatus::UpdateIntervalEnd;
+            } else {
+              intervalStatus_ = IntervalStatus::SameInterval;
             }
-            cacheReset();
           }
         }
       }
-      return returnValue;
+      return intervalStatus_ != IntervalStatus::Invalid;
```

Comment:
What is driving the need for `intervalStatus_`? I'm not really thrilled with requiring the class to be so modal.

Reply:
After some verbal discussions with Chris, we decided the code in this PR was this best way to implement this for now, although we may revisit it later in a later PR if we have new ideas. The intervalStatus_ serves two purposes. First, it prevents executing the work in setValidityIntervalFor multiple times for the same sync value. Second it saves the result of a calculation that determines whether we need to start a new IOV or update the end of the IOV or do nothing or have an invalid IOV.